integration-test/src/test/java/org/sagebionetworks/warehouse/WarehouseTestHelperImpl.java

```diff
@@ -0,0 +1,173 @@
+import org.sagebionetworks.StackConfiguration;
+import org.sagebionetworks.common.util.Clock;
+import org.sagebionetworks.util.Pair;
+import org.sagebionetworks.util.TimeUtils;
+
+import com.amazonaws.services.athena.AmazonAthena;
+import com.amazonaws.services.athena.model.GetQueryExecutionRequest;
+import com.amazonaws.services.athena.model.GetQueryResultsRequest;
+import com.amazonaws.services.athena.model.QueryExecution;
+import com.amazonaws.services.athena.model.QueryExecutionContext;
+import com.amazonaws.services.athena.model.QueryExecutionState;
+import com.amazonaws.services.athena.model.ResultSet;
+import com.amazonaws.services.athena.model.StartQueryExecutionRequest;
+import com.amazonaws.services.s3.AmazonS3;
+
+public class WarehouseTestHelperImpl implements WarehouseTestHelper {
+
+	public static final int MAX_WAIT_MS = 60_000;
+	public static final int WAIT_INTERAVAL_MS = 1000;
+	public static final String BUCKET_NAME = "dev.testdata.sagebase.org";
+
+	private final AmazonS3 s3Client;
+	private final AmazonAthena athenaClient;
+	private final Clock clock;
+	private final StackConfiguration stackConfig;
+
+	public WarehouseTestHelperImpl(AmazonS3 s3Client, AmazonAthena athenaClient, Clock clock,
+			StackConfiguration stackConfig) {
+		super();
+		this.s3Client = s3Client;
+		this.athenaClient = athenaClient;
+		this.clock = clock;
+		this.stackConfig = stackConfig;
+	}
+
+	@Override
+	public void assertWarehouseQuery(String queryString, int maxNumberOfHours) throws Exception {
+
+		Instant now = Instant.ofEpochMilli(clock.currentTimeMillis());
+
+		StackTraceElement callersElement = Thread.currentThread().getStackTrace()[2];
+		String callersPath = getCallersPath(callersElement);
+
+		saveQueryToS3(queryString, now, callersPath, maxNumberOfHours);
+
+		List<String> previousQueryKeysToCheck = s3Client.listObjectsV2(BUCKET_NAME, callersPath).getObjectSummaries()
+				.stream().filter(s -> now.toEpochMilli() >= getExpiresOnFromKey(s.getKey())).map(s -> s.getKey())
+				.collect(Collectors.toList());
+		
+		try {
+			assertAll("Group of all previous queries ready to be tested",
+					previousQueryKeysToCheck.stream().map(this::executeQueryAndAssertResults));
+		} finally {
+			previousQueryKeysToCheck.forEach(key -> s3Client.deleteObject(BUCKET_NAME, key));
+		}
+	}
+
+	String getCallersPath(StackTraceElement callersElement) {
+		String callersPath = String.format("%s/%s/%s/%d", stackConfig.getStackInstance(),
+				callersElement.getClassName().replaceAll("\\.", "/"), callersElement.getMethodName(),
+				callersElement.getLineNumber());
+		return callersPath;
+	}
+	
+	Executable executeQueryAndAssertResults(String key) {
+		return ()->{
+			String previousQueryString = s3Client.getObjectAsString(BUCKET_NAME, key);
+			int count = executeCountQuery(previousQueryString);
+			assertTrue(count > 0, "Expected count > 0 for query: " + previousQueryString);
+		};
+	}
+	
+	int executeCountQuery(String previousQueryString) throws Exception {
```

Comment:
What if you try to compact all the queries into one with a UNION ALL? E.g.
select count(*) from blablabla
UNION ALL
select count(*) from blablabla